docs/architecture/adr-008-app-caller-cbs.md

```diff
@@ -0,0 +1,111 @@
+# ADR 008: Callback to IBC-App Callers
+
+## Changelog
+* 2022-08-10: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+IBC was designed with callbacks between core IBC and IBC applications. IBC apps would send a packet to core IBC. When the result of the packet lifecycle eventually resolved into either an acknowledgement or a timeout, the core IBC called a callback on the IBC application so that the IBC application could take action on the basis of the result (e.g. unescrow tokens for ICS-20)
+
+We are now seeing the desire for secondary applications to call into IBC apps as part of their state machine logic and then do some actions on the basis of the packet result.
+
+E.g. Send an ICS-20 packet, and if it is successful, then send an ICA-packet to swap tokens on LP and return funds to sender
+
+This requires a second layer of callbacks. The IBC application already gets the result of the packet from core IBC, but currently there is no standardized way to pass this information on to a caller module/smart contract.
+
+## Decision
+
+Create a standardized callback interface that callers can implement. IBC applications (or middleware that wraps IBC applications) can now call this callback to route the result of the packet/channel handshake from core IBC to the IBC application to the original caller.
+
+IBC callers may implement the following interface:
+
+```go
+type IBCCaller interface {
```

Comment:
Should we add and `error` return type to these functions in case something happens?

Reply:
Can we also have `OnRecvPacket`? 

One use I can think of which I might want is..
User ibc-transfers atom from cosmoshub to persistence chain - without xprt on the origin wallet the user cannot do anything with the atom.
If there were a callback that uses some sort of swap on persistence for transaction fees worth of xprt OnRecvPacket, that could make UX easier

Reply:
Agreed that this would be useful! It's not quite in scope for this ADR since the receiver address is not the caller of the IBC application.

Also, there is no generic definition of receiver address for IBC packets since they may not include one

Reply:
> User ibc-transfers atom from cosmoshub to persistence chain - without xprt on the origin wallet the user cannot do anything with the atom.
If there were a callback that uses some sort of swap on persistence for transaction fees worth of xprt OnRecvPacket, that could make UX easier

This usecase in particular does not require this ADR. Please see [middleware guide](https://github.com/cosmos/ibc-go/blob/main/docs/ibc/middleware/develop.md)

Reply:
I changed the semantics of this ADR so that it is back in scope and specified in this ADR. Thanks for the feedback @puneet2019 !